store/store.go

```diff
@@ -114,27 +131,48 @@ func NewStore(base string) (*Store, error) {
 					return err
 				}
 			}
-		}
-		// if db is populated check its version
-		version, err := getDBVersion(tx)
-		if err != nil {
-			return err
-		}
-		if version < dbVersion {
-			// TODO(sgotti) execute migration functions
-			return fmt.Errorf("Current cas db version: %d lesser than the current rkt expected version: %d", version, dbVersion)
-		}
-		if version > dbVersion {
-			return fmt.Errorf("Current cas db version: %d greater than the current rkt expected version: %d", version, dbVersion)
+		} else {
+			// if db is populated check its version
+			version, err := getDBVersion(tx)
+			if err != nil {
+				return err
+			}
+			if version < dbVersion {
+				needsMigrate = true
+			}
+			if version > dbVersion {
+				return fmt.Errorf("Current store db version: %d greater than the current rkt expected version: %d", version, dbVersion)
+			}
 		}
 		return nil
 	}
-
 	if err = db.Do(fn); err != nil {
 		return nil, err
 	}
 
-	ds.treestore = &TreeStore{path: filepath.Join(base, "cas", "tree")}
+	// migration is done in another transaction as it must take an exclusive
+	// store lock. If, in the meantime, another process has already done the
+	// migration, between the previous db version check and the below
+	// migration code, the migration will do nothing as it'll start
+	// migration from the current version.
+	if needsMigrate {
+		// Take an exclusive store lock
+		err := ds.storeLock.ExclusiveLock()
+		if err != nil {
+			return nil, err
+		}
+		// TODO(sgotti) take a db backup (for debugging and last resort rollback?)
```

Comment:
Yeah, this would be prudent (although we can do in follow up if not part of this review)